_posts/2019-04-10-aws-dashboard-console-using-arduino.md

```diff
@@ -0,0 +1,265 @@
+
+**After the LED’s are fine, write content to the LCD:**
+
+```c++
+  for (int i=0; i < 5; i++) {
+
+    bool all_ok = true;
+  
+    if (pipelines_failed_dev == true) {
+      lcd.setCursor(0,0);
+      lcd.print("DEV Pipe fail:  ");
+      lcd.setCursor(0,1);
+      lcd.print(pipes_failed_0_dev);
+
+      all_ok = false;
+      delay(5000);
+    } 
+    
+    if (alarms_raised_prod == true) {  
+      lcd.setCursor(0,0);
+      lcd.print("PROD Alarms:    ");
+      lcd.setCursor(0,1);
+      lcd.print(alarms_0_prod);
+
+      all_ok = false;
+      delay(5000);
+    }
+
+    ...Other clauses...
+
+    if (all_ok == true) {  
+      lcd.clear();
+      lcd.print("     All OK     ");
+
+      delay(60000);
+
+      break;
+    }
+  }
+```
+
+Every `if` clause contains 5 second delay to keep the certain text on the screen for a moment. The `for` loop is used to “blink” all texts few times on the screen. If there are nothing to show, then print “All OK”.
+
+Few examples of the LCD output:
+
+![Dashboard LCD](/img/aws-dashboard-console-using-arduino/dashboard-lcd.png "Dashboard LCD")
+
+If the string is longer than 16 characters (LCD width), then it is cut. 
+
+>*The LCD library has an extra option to scroll overflowing content on the screen but I note that it was not suitable for this case. Scrolling every string on the screen would take just too much time.*
+
+### Lambda backend script
+AWS status information is served by the simple Lambda script (Python). The Lambda provides a REST interface that includes all necessary information for the console. The Lambda starts on request and does not contain persistence. The Lambda is publised though the AWS API Gateway and secured by an API KEY.
+
+An example of the Lambda response:
+
+```json
+{
+    "alarms_list": [],
+    "alarms_raised": false,
+    "alarms_raised_history": false,
+    "pipelines_running": false,
+    "pipelines_running_list": [],
+    "pipelines_failed": false,
+    "pipelines_failed_list": []
+}
+```
+
+## Summary
+
+The project contained few stages:
+1. Planning the required components
+2. Prototyping a first circuit version using breadboard
+3. Coding the program
+4. Building a final version to the box and solder the required wiring
+
+The most laborious stage was building the box and place all components to it. I found the plastic box from the local electronic shop and made the required holes by a drill and knife. That took a surprising amount of time and the result is still a bit rough.
+
+The coding was relatively easy but there was still some troubles on the road. Originally I used an old version of the JSON parser that affected issues. The old version was using C++ pointers that I had never used before. The pointers affected a strange behaviour when the JSON properties was print to the LCD. After updating a latest version of the JSON parser, everything worked fine.
```

Comment:
JSON parser that **caused** issues
The pointers **caused**
properties **were** print**ed**